### ISSUE TITLE
Integrate SROP into ROP

### ISSUE THREAD

**Issue by zachriggle** [binjitsu/binjitsu#11]
I think we'll want to integrate this a bit more into the workflow for the current `ROP` module and class.

Currently, we hook `__getattr__` of the `ROP` objects so that you can do something like:

```
r = ROP([binary, libc])
r.read(0, 0xdeadbeef, 32) 
```

And it will automatically resolve the location of `read` in the PLT of the main executable, or libc.

We'd like to have this automatically occur for things which are system calls (i.e. `read` but not `system`) by leveraging the SROP functionality.  The user of this functionality should not need to know that they're using SROP, it should just kind of happen.

See #9

**Comment by eQu1NoX** [comment]
Hi,

Just to be clear, is the idea to have something along the lines of 
s = srop([binary, libc])
s.mprotect(address, size, protection)

and internally it would set the eip to an "int 0x80/syscall" instruction and invoke the mprotect system call? This would be nice to have indeed.

We do need to think of how we control esp/rsp though; perhaps expect that from the user too? Does that sound okay?

**Comment by zachriggle** [comment]
Nope, the idea is exactly as shown above.  In the event that `read` is not an exported symbol in any of the available libraries (e.g. if `libc` is not provided) but a syscall gadget *is* available, it should transparently switch to SROP without the user knowing.

**Comment by zachriggle** [comment]
Also, the idea is to allow mixing of normal ROP and SROP transparently.  The place to catch all of this is in `pwnlib.rop.ROP.call`, specifically the case where the function cannot be resolved.  If we can locate a syscall gadget (or SVC on ARM, etc), then we would do it there.

We'll probably also need to extend the representation of `pwnlib.rop.ROP.dump` in order to display which register values correspond to the entries on the stack.  For example:

```
$ echo 'int main(){}' | gcc -m32 -xc -
```

```
>>> aout=ELF('a.out')
>>> libc=ELF('/lib/i386-linux-gnu/libc.so.6')
>>> rop=ROP([aout,libc])
>>> rop.read(0, 0x1234000, 1024)
```

When `read` is available, you get the following:

```
>>> print r.dump()
0x0000:          0xdabd0 (__read)
0x0004:       0xdeadbeef
0x0008:              0x0
0x000c:        0x1234000
0x0010:            0x400
```

When using SROP, it might do something like below.  Obviously there are some trash registers that don't matter, so we can probably skip them, but it's useful to show which registers we know about that matter.

```
print r.dump()
0x0000:          0xe6dc0 (syscall)
0x0004:       0xdeadbeef
0x0008:              0x3 (eax)
0x000c:              0x0 (edi)
0x0010:        0x1234000 (esi)
0x0014:            0x400 (edx)
```

**Comment by eQu1NoX** [comment]
Hmm I see. However, when we make an SROP call we do end up loading all registers. As a result, if we do not specify a value for ESP, it will simply be loaded with 0x0. I'm not sure if we can transparently switch to SROP due to the fact that ESP needs to be loaded with some user supplied value when any srop call is made. From what I understand there is no way of making a SROP call to load up registers and say "dont touch esp, leave it as it is".
Hmm, does that make sense?

**Comment by zachriggle** [comment]
That's true.  We can do it transparently as long as it's the *last* gadget,
or if we know where the ROP chain is (each ROP object has a `.base`
attribute that we use for self-pointers).

We can require .base is specified, or that it's the last gadget.


*Zach Riggle*

**Comment by eQu1NoX** [comment]
Got it. I'll check out pwnlib.rop.ROP.call and pwnlib.rop.ROP.dump.

**Comment by eQu1NoX** [comment]
Actually, I was thinking about this a bit more. Its not just sufficient to have a syscall gadget, we need a sigreturn gadget. The sigreturn gadget would be something that moved the system call number of sigreturn into eax/rax and then executed the "int 80/syscall" instruction.
i.e. on linux32 bit the sigreturn gadget would be "mov eax, 0x77; int 0x80; ret".

So at the top of the frame we would have the address of the sigreturn gadget, and following that we would have the sigreturn frame that invoked the required system call. We would do this by loading up the registers with the required arguments, and having eip point to a syscall instruction.

Currently, I'm looking for whether we have sigreturn gadgets inside libc. If not, I'll try and see if rop.search can be used to find such a gadget.

**Comment by zachriggle** [comment]
Is there any reason a (E|R)AX-controlling gadget would be insufficient?

**Comment by zachriggle** [comment]
For example:

```
>>> r=ROP('/lib/i386-linux-gnu/libc.so.6')
[*] '/lib/i386-linux-gnu/libc.so.6'
    Arch:     i386-32-little
    RELRO:    Partial RELRO
    Stack:    Canary found
    NX:       NX enabled
    PIE:      PIE enabled
[*] Loaded cached gadgets for '/lib/i386-linux-gnu/libc.so.6'
>>> r.eax
(149151L, {'insns': [u'pop eax', u'ret'], 'move': 8, 'regs': [u'eax']})
```

**Comment by eQu1NoX** [comment]
Ah that'd be perfect, thanks!

**Comment by zachriggle** [comment]
Closed by #16
